x/bundles/keeper/logic_bundles.go

```diff
@@ -245,37 +244,61 @@ func (k Keeper) calculatePayouts(ctx sdk.Context, poolId uint64, totalPayout uin
 		return
 	}
 
+	// Already return if there are no payouts
+	if totalPayout.IsZero() {
+		return
+	}
+
 	bundleReward.Total = totalPayout
 
 	// calculate share of treasury from total payout
-	bundleReward.Treasury = uint64(math.LegacyNewDec(int64(totalPayout)).Mul(k.GetNetworkFee(ctx)).TruncateInt64())
+	treasuryPayout, _ := sdk.NewDecCoinsFromCoins(totalPayout...).MulDec(k.GetNetworkFee(ctx)).TruncateDecimal()
+	bundleReward.Treasury = treasuryPayout
+
+	// subtract treasury payout from total payout, so we can continue splitting up the rewards from here
+	totalPayout = totalPayout.Sub(bundleReward.Treasury...)
+	if totalPayout.IsZero() {
+		return
+	}
+
+	// subtract storage cost from remaining total payout. We split the storage cost between all coins and charge
+	// the amount per coin. If there is not enough of that coin available we simply charge what is left, so there
+	// can be the case that the storageRewards are less than what we actually wanted to pay out. This is acceptable
+	// because this case is very rare, usually the minFundingAmount ensures that there are always enough funds left
+	// of each coin, and in the case there are not enough the coins are removed and therefore for the next bundle
+	// we split between the other remaining coins.
+	whitelist := k.fundersKeeper.GetCoinWhitelistMap(ctx)
+	wantedStorageRewards := sdk.NewCoins()
+	storageCostPerCoin := k.GetStorageCost(ctx, bundleProposal.StorageProviderId).MulInt64(int64(bundleProposal.DataSize)).QuoInt64(int64(totalPayout.Len()))
+
+	for _, coin := range totalPayout {
+		amount := storageCostPerCoin.Mul(whitelist[coin.Denom].CoinWeight)
```

Comment:
Lets clarify the meaning and unit of CoinWeight. 
It makes sense to use
[storageCost] = USD/{byte}
[coinWeight] = USD/{denom}
The comment should als be added to the proto file then.

coinWeight != 0:
then it would be storageCostPerCoin / coinWeight